src/hotspot/share/runtime/continuationFreezeThaw.cpp

```diff
@@ -1047,7 +1047,7 @@ NOINLINE freeze_result FreezeBase::recurse_freeze_interpreted_frame(frame& f, fr
   // The frame's top never includes the stack arguments to the callee
   intptr_t* const stack_frame_top = ContinuationHelper::InterpretedFrame::frame_top(f, callee_argsize, callee_interpreted);
   const int locals = f.interpreter_frame_method()->max_locals();
-  const int fsize = f.fp() + frame::metadata_words + locals - stack_frame_top;
+  const int fsize = f.fp() NOT_RISCV64(+ frame::metadata_words) + locals - stack_frame_top;
 
   intptr_t* const stack_frame_bottom = ContinuationHelper::InterpretedFrame::frame_bottom(f);
   assert(stack_frame_bottom - stack_frame_top >= fsize, ""); // == on x86
```

Comment:
Are there any ports where (stack_frame_bottom - stack_frame_top) != fsize?  It would be nice if we could use (stack_frame_bottom - stack_frame_top) for fsize and remove the platform-specific computation using frame::metadata_words above.

Reply:
I think aarch64 and riscv is different from x86 here due to possible padding in the frame[1].
So if we modify this assertion like:
```
diff --git a/src/hotspot/share/runtime/continuationFreezeThaw.cpp b/src/hotspot/share/runtime/continuationFreezeThaw.cpp
index 2ef48618ccb..c8e88b67f94 100644
--- a/src/hotspot/share/runtime/continuationFreezeThaw.cpp
+++ b/src/hotspot/share/runtime/continuationFreezeThaw.cpp
@@ -1050,7 +1050,7 @@ NOINLINE freeze_result FreezeBase::recurse_freeze_interpreted_frame(frame& f, fr
   const int fsize = f.fp() + frame::metadata_words + locals - stack_frame_top;

   intptr_t* const stack_frame_bottom = ContinuationHelper::InterpretedFrame::frame_bottom(f);
-  assert(stack_frame_bottom - stack_frame_top >= fsize, ""); // == on x86
+  assert(stack_frame_bottom - stack_frame_top == fsize, ""); // == on x86

   DEBUG_ONLY(verify_frame_top(f, stack_frame_top));
```

Then we will trigger assertion failure on linux-aarch64:
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/home/realfyang/openjdk-jdk/src/hotspot/share/runtime/continuationFreezeThaw.cpp
:1053), pid=2680946, tid=2680964
#  Error: assert(stack_frame_bottom - stack_frame_top == fsize) failed
#
# JRE version: OpenJDK Runtime Environment (20.0) (slowdebug build 20-internal-adhoc.realfyang.open
jdk-jdk)
# Java VM: OpenJDK 64-Bit Server VM (slowdebug 20-internal-adhoc.realfyang.openjdk-jdk, mixed mode,
 sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-aarch64)
# Problematic frame:
# V  [libjvm.so+0x865c98]  FreezeBase::recurse_freeze_interpreted_frame(frame&, frame&, int, bool)+
0xb8
#
```

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/frame_aarch64.hpp#L62

Reply:
OK, skipping the assert for now.  Regarding `NOT_RISCV64(+ frame::metadata_words)` @reinrich just posted the PR for the PPC64 port (see JDK-8286302), which introduces metadata_words_at_top/metadata_words_at_bottom.  I'm not sure what values RISC-V would use for these new constants, but merging with the PPC64 changes might allow a platform-independent setting of fsize here.

Reply:
Hi, I went through the PPC64 changes to shared code and I think for RISC-V we should define metadata_words_at_top and metadata_words_at_bottom to 0 and 2 (which equals metadata_words as defined by this PR) respectively. The PPC64 changes will help eliminate the RISC-V-specific change made in stackChunkOopDesc::is_usable_in_chunk in file src/hotspot/share/oops/stackChunkOop.inline.hpp. But we might still need a RISC-V platform-dependent change here due to the particularity of the RISC-V frame structure. Thanks.

Reply:
The value for `fsize` as it is calculated is correct on PPC64 but only because errors compensate. The calculation makes assumptions about the position of fp which is not specified.

This is the assumed layout:

```
  :                 :
  :                 :
  |                 |
  |-----------------|
  |                 |
  | locals array    |
  |                 |<- callers_SP
  ===================
  |                 |
  | metadata@bottom |
  |                 |<- FP
  |-----------------|
  |                 |
  |                 |
  |                 |
  |                 |
  |                 |<- SP
  ===================
```

Here the metadata lies outside [FP, SP] and needs to be added to FP-SP just like the size of the locals array.

This coincidentally matches the layout on PPC64:

```
  :                 :
  :                 :
  |                 |
  |-----------------|
  |                 |
  | locals array    |
  |                 |
  |-----------------|
  |                 |
  | metadata@top    |
  |                 |<- FP / callers_SP
  ===================
  |                 |
  |                 |
  |                 |
  |                 |
  |-----------------|
  |                 |
  | metadata@top    |
  |                 |<- SP
  ===================

```

I assume the layout on Risc-V is like this:

```
  :                 :
  :                 :
  |                 |
  |-----------------|
  |                 |
  | locals array    |
  |                 |<- FP / callers_SP
  ===================
  |                 |
  | metadata@bottom |
  |                 |
  |-----------------|
  |                 |
  |                 |
  |                 |
  |                 |
  |                 |<- SP
  ===================
```
Here the metadata is included in [FP, SP]


We could change the line to

```c++
  const int fsize = ContinuationHelper::InterpretedFrame::callers_sp(f) + frame::metadata_words_at_top + locals - stack_frame_top;
```

with a platform dependent implementation of `callers_sp()`

Reply:
@reinrich : Yes, you are right about the frame layout of RISC-V. And your proposed solution looks reasonable to me. Since that will need testing accross all platforms, maybe we can do that as a separate PR later on after this is merged. The RISCV testing platform is relatively slow and this PR has been tested for a long time. Thanks.

Reply:
@RealFYang I'm ok with leaving this for later. The change is actually rather small. I've added it to the PPC64 port with https://github.com/openjdk/jdk/pull/10961/commits/0d12b0577d8314ed1d571d9cac65fa8866939180
Quick tests succeeded on X86, AARCH64 and PPC64.

Reply:
@reinrich : That's great to hear. Thank you. After this PR is merged, I can help test your changes on RISC-V after you rebase.